stdlib/LinearAlgebra/src/LinearAlgebra.jl

```diff
@@ -388,6 +388,126 @@ const ⋅ = dot
 const × = cross
 export ⋅, ×
 
+"""
+    hadamard(a, b)
+    a ⊙ b
+
+Elementwise multiplication of `a` and `b`. This yields the same result as `map(*, a, b)`
+if `a` and `b` have the same axes, and throws an error otherwise.
+
+`⊙` can be passed as an operator to higher-order functions.
+
+```jldoctest
+julia> a = [2, 3]; b = [5, 7];
+
+julia> a ⊙ b
+2-element Array{$Int,1}:
+ 10
+ 21
+
+julia> a ⊙ [5]
+ERROR: DimensionMismatch("in the axes of `A` and `B` must match, got (Base.OneTo(2),) and (Base.OneTo(1),)")
+[...]
+```
+
+!!! compat "Julia 1.5"
+    This function requires at least Julia 1.5. In Julia 1.0-1.4 it is available from
+    the `Compat` package.
+"""
+function hadamard(A, B)
+    @noinline throw_dmm(axA, axB) = throw(DimensionMismatch("in the axes of `A` and `B` must match, got $axA and $axB"))
+
+    axA, axB = axes(A), axes(B)
+    axA == axB || throw_dmm(axA, axB)
+    return map(*, A, B)
+end
+const ⊙ = hadamard
+
+"""
+    hadamard!(dest, A, B)
+
+Similar to `hadamard(A, B)` (which can also be written `A ⊙ B`), but stores its results in
+the pre-allocated array `dest`.
+
+!!! compat "Julia 1.5"
+    This function requires at least Julia 1.5. In Julia 1.0-1.4 it is available from
+    the `Compat` package.
+"""
+function hadamard!(dest::AbstractArray, A::AbstractArray, B::AbstractArray)
+    @noinline function throw_dmm(axA, axB, axdest)
+        throw(DimensionMismatch("`axes(dest) = $axdest` must be equal to `axes(A) = $axA` and `axes(B) = $axB`"))
+    end
+
+    axA, axB, axdest = axes(A), axes(B), axes(dest)
+    ((axdest == axA) & (axdest == axB)) || throw_dmm(axA, axB, axdest)
+    @simd for I in CartesianIndices(axdest)
+        @inbounds dest[I] = A[I]*B[I]
+    end
+    return dest
+end
+
+export ⊙, hadamard, hadamard!
+
+"""
+    tensor(A, B)
+    A ⊗ B
+
+Compute the tensor product of `A` and `B`.
+If `C = A ⊗ B`, then `C[i1, ..., im, j1, ..., jn] = A[i1, ... im] * B[j1, ..., jn]`.
+
+```jldoctest
+julia> a = [2, 3]; b = [5, 7, 11];
+
+julia> a ⊗ b
+2×3 Array{$Int,2}:
+ 10  14  22
+ 15  21  33
+```
+
+See also: [`kron`](@ref).
+
+!!! compat "Julia 1.5"
+    This function requires at least Julia 1.5. In Julia 1.0-1.4 it is available from
+    the `Compat` package.
+"""
+tensor(A::AbstractArray, B::AbstractArray) = [a*b for a in A, b in B]
+const ⊗ = tensor
+
+function tensor(u::AbstractVector, v::Adjoint{T,<:AbstractVector}) where T
+    error("`u ⊗ v'` is not defined, perhaps you meant `u * v'` or `u * transpose(v)`")
+end
+function tensor(u::AbstractVector, v::Transpose{T,<:AbstractVector}) where T
+    error("`u ⊗ v'` is not defined, perhaps you meant `u * v'` or `u * transpose(v)`")
+end
```

Comment:
It seems pretty arbitrary to me to error here, because `Adjoint` and `Transpose` are `AbstractMatrices` as well. It might be good to document this as a gotcha, but I would be quite confused if `[1, 2, 3]'` didn't work while `[1 2 3]` did.

Reply:
See the rollicking discussion above, particularly https://github.com/JuliaLang/julia/pull/35150#issuecomment-602837207 and https://github.com/JuliaLang/julia/pull/35150#issuecomment-602549570. IMO it's so ambiguous it's better to error.

Reply:
Maybe @Jutho could comment on this, but my understanding is that if we clearly define `⊗` to be a tensor product, it wouldn't be inconsistent. The problem only comes in if you want to differentiate between covariant and contravariant indices. We don't currently have a way to really do express that, so I think it makes sense to treat these as matrices, not covectors, as we already do in other places.

Reply:
Theoretically, there is indeed no issue with the tensor product of a vector and a dual vector. Practically, I am not opposed to this error in order to avoid confusion with the outer product .

Reply:
I hope I'm not just adding noise here (🙏), but my inclination would be to let  `u ⊗ v'` be a matrix (because it is), but I don't know what the challenges to implementation might be.

Matrices are once-contravariant and once-covariant tensors and matrix multiplication is actually partial evaluation:

```julia
M = sum_{i,j} M^i_j e_i ⊗ dx^j

X = sum_k X^k e_k

M(X) = sum_{i,j,k} M^i_j e_i X^k dx^j(e_k) = sum_{i,j} M^i_j X^j e_i
```

where `e_i` is a basis vector and `dx^i = (e_i)'` is a basis covector.

I guess you would define:

```julia
M*X := M(X).
```

To do all this correctly requires more thought I think, but may be outside the scope of this PR, which I wouldn't object if you want to go ahead.

Reply:
I don't think that line of thinking fits well within Julia's current type hierarchy though. `Adjoint` and `Transpose` are both `<:AbstractMatrix`, and also behave like a matrix when indexing, for example. This also throws up other questions, like what the tensor product of a vector and vector or covector and covector should be. You say, matrices are (1,1) tensors, but these operations would produce (2,0), or (0,2) respectively, tensors, do we still want to use normal matrices for that? It is also ambiguous for higher dimensional arrays, which indices are raised or lowered. Since we handle transpose specially, shouldn't we also ignore (leading) dimensions of size 1 from other arrays?
I very much would like a better interface for arrays as tensors in Julia, but I think this needs to be supported everywhere and we should be consistent with other parts of base here. The problem I have with throwing an error instead, is that this makes it a lot harder to write generic code, because in Julia, you can pretty much always assumes that a function that works for AbstractMatrices also just works for `Transpose{<:Vector}` and forcing users to special case `Adjoint` and `Transpose` in generic code, seems unnecessarily convoluted for most use cases.

Reply:
But an `Adjoint` and an `AbstractMatrix` behave differently even with respect to regular multiplication, i.e. compare:
```julia
a = randn(3)
x = a' * a 
y = Matrix(a') * a
```

Reply:
Yes, but you can still index scalars just like you would a matrix, and all arithmetic operations should behave the same on a 1x1 matrix as on a scalar. Inserting / removing dimensions from in between arrays is different, because indexing no longer works the same and you would always have to be careful, which dimension of the output now corresponds to which dimension of the input arrays.

Reply:
Hence, I think that erroring is probably the best way forward for this specific operation: `tensor(u, v')`. There will be people that argue that it needs to be a rank two tensor, because `v'` is a dual vector but only has one actual index/dimension, and there will be people that argue that it should be a rank three tensor, because `v'` behaves like a `Matrix` and has two dimensions.

Reply:
Ok, since people have a lot of different opinions on this, I am also inclined to say that throwing an error is ok for now. We can always make this not an error later on.

Reply:
My vote would be that `u ⊗ v'` is a matrix -- this is within LinearAlgebra, where `v'` is really a vector with a tag saying that its index is downstairs. But an error is OK too, and leaves the door open. 

Should `v' ⊗ u` also be an error? It's hard to picture that getting a size `(1, N, M)` array is what's desired. And `M ⊗ v'` with a matrix or higher?

Reply:
Is there a good place to discuss the broader issues that I think still need some thought that go beyond this PR? I don't want to distract, but think some more discussion is probably warranted (maybe for v2.0).

Reply:
Sounds like another "Taking...seriously" issue. (Though methinks we need a new meme...)

Reply:
Indeed @mcabbott , given the current errors, `tensor` should error if either of its arguments is a `Transpose{_,<:AbstractVector}` or `Adjoint{_,<:AbstractVector}`, so these types (i.e. dual vectors / covectors) just don't get to play along with the new `tensor` product.